### PR TITLE
Support Ruby 3.1

### PR DESCRIPTION
**Pull request by a2ikm**
This PR adds Ruby 3.1 support.

### CODE REVIEW

File: gemfiles/Gemfile.common.rb [hunk 1]
```diff
@@ -11,7 +11,7 @@
 group :test do
   gem 'test-unit',      '~> 3.0'
   gem 'rspec',          '~> 3.4.0'
-  gem 'timecop',        '= 0.6.1'
+  gem 'timecop',        '>= 0.9.4'
```
Comment (a2ikm) [inline gemfiles/Gemfile.common.rb:14]: We need timecop 0.9.4+ for ruby 3.1+.
https://github.com/travisjeffery/timecop/pull/288

File: lib/mongo_mapper/plugins.rb [hunk 2]
```diff
@@ -14,9 +22,18 @@ def plugin(mod)
       plugins << mod
     end
 
+    def direct_descendants
+      Plugins.direct_descendants_hash[self]
+    end
```
Comment (a2ikm) [inline lib/mongo_mapper/plugins.rb:27]: Given Ruby 3.0 and ActiveSupport 7.0, ActiveSupport::DescendantsTracker#direct_descendants doesn't work with Modules. So let's implement it by ourselves.

File: spec/unit/document_spec.rb [hunk 3]
```diff
@@ -88,7 +88,7 @@ class Post
     end
 
     it "should be recorded" do
-      Message.descendants.should == [Enter, Exit, Chat]
+      Message.descendants.should contain_exactly(Enter, Exit, Chat)
```
Comment (a2ikm) [inline spec/unit/document_spec.rb:91]: Our own `direct_descendants` implementation changes the order of `descendants` from the previous implementation. I think it doesn't matter.

**Comment by smtlaissezfaire** [comment]
Will review tonight or tomorrow.

File: spec/unit/document_spec.rb (hunk 3 above)
Comment (smtlaissezfaire) [reply spec/unit/document_spec.rb:91]: LGTM, don't think the order is important.

File: gemfiles/Gemfile.common.rb (hunk 1 above)
Comment (smtlaissezfaire) [reply gemfiles/Gemfile.common.rb:14]: Only a test dependency so think this has no worries.

File: lib/mongo_mapper/plugins.rb (hunk 2 above)
Comment (smtlaissezfaire) [reply lib/mongo_mapper/plugins.rb:27]: Looks good

**Review by smtlaissezfaire — COMMENTED** [review]
Made a small change, couldn't push to your branch so pushed to mongomapper/ruby-3-1 branch (https://github.com/mongomapper/mongomapper/tree/ruby-3-1).

Noticed a whole bunch of warnings:
```
Calling `DidYouMean::SPELL_CHECKERS.merge!(error_name => spell_checker)' has been deprecated. Please call `DidYouMean.correct_error(error_name, spell_checker)' instead.
/Users/smtlaissezfaire/.rvm/rubies/ruby-3.1.2/bin/ruby -I/Users/smtlaissezfaire/.rvm/gems/ruby-3.1.2/gems/rspec-core-3.4.4/lib:/Users/smtlaissezfaire/.rvm/gems/ruby-3.1.2/gems/rspec-support-3.4.1/lib /Users/smtlaissezfaire/.rvm/gems/ruby-3.1.2/gems/rspec-core-3.4.4/exe/rspec --pattern spec/\*\*/\*_spec.rb --color
Calling `DidYouMean::SPELL_CHECKERS.merge!(error_name => spell_checker)' has been deprecated. Please call `DidYouMean.correct_error(error_name, spell_checker)' instead.
/Users/smtlaissezfaire/.rvm/gems/ruby-3.1.2/gems/rspec-core-3.4.4/lib/rspec/core/configuration_options.rb:168: warning: Passing safe_level with the 2nd argument of ERB.new is deprecated. Do not use it, and specify other arguments as keyword arguments.
/Users/smtlaissezfaire/.rvm/gems/ruby-3.1.2/gems/rspec-core-3.4.4/lib/rspec/core/configuration_options.rb:168: warning: Passing trim_mode with the 3rd argument of ERB.new is deprecated. Use keyword argument like ERB.new(str, trim_mode: ...) instead.
Calling `DidYouMean::SPELL_CHECKERS.merge!(error_name => spell_checker)' has been deprecated. Please call `DidYouMean.correct_error(error_name, spell_checker)' instead.
```

Any idea where these are coming from?

Also, I noticed I had to `bundle update` (`bundle install` didn't work, failing on a nokogiri version that wasn't supported on ruby 3.1.).  Not sure if that was something with my new installation of ruby 3.1.  Seems to work on the build server though, so I'm inclined to think it's not an issue.

**Comment by a2ikm** [comment]
@smtlaissezfaire
Thanks for your review!

I cherry-picked your additional commit to this branch.
I will create a new branch in mongomapper/mongomapper from the next time :)

> Calling `DidYouMean::SPELL_CHECKERS.merge!(error_name => spell_checker)' has been deprecated. Please call `DidYouMean.correct_error(error_name, spell_checker)' instead.

Oh, I've overlooked it.
This warning seems for bundler and thor.
https://github.com/rubygems/rubygems/issues/5234
https://github.com/rails/thor/pull/761

Running `bundle update --bundler` and `bundle update thor` soved it on my machine.

> Also, I noticed I had to `bundle update` 

I've updated timecop and it requires `bundle update`.
https://github.com/mongomapper/mongomapper/pull/692/files#diff-d5a6305427ceeeda936968b8681db00a82d9d7af9120bd838a12b0793aa1c5fa.
But I didn't have any trouble for nokogiri.

**Comment by smtlaissezfaire** [comment]
@a2ikm Why doesn't `bundle update` update the Gemfile.lock?

**Comment by smtlaissezfaire** [comment]
Oh, it's `.gitignore`'d.  I don't think it should be?

Going to merge this.

**Pull request merged**